cpp/include/raft/core/detail/copy.hpp

```diff
@@ -0,0 +1,510 @@
+{
+  return apply(std::forward<LambdaT>(lambda), std::forward<ContainerT>(args),
+make_index_sequence<IdxT, size>{});
+} */
+
+/*
+ * Given an mdspan and an array of indices, return a reference to the
+ * indicated element.
+ */
+template <typename MdspanType, typename IdxType, IdxType... Idx>
+__device__ auto& get_mdspan_elem(MdspanType& md,
+                                 IdxType const* indices,
+                                 index_sequence<IdxType, Idx...>)
+{
+  return md(indices[Idx]...);
+}
+
+template <typename MdspanType, typename IdxType>
+__device__ auto& get_mdspan_elem(MdspanType& md, IdxType const* indices)
+{
+  return get_mdspan_elem(
+    md, indices, make_index_sequence<IdxType, MdspanType::extents_type::rank()>{});
+}
+
+/* Advance old_indices forward by the number of mdspan elements specified
+ * by increment. Store the result in indices. Return true if the new
+ * indices are valid for the input mdspan.
+ */
+template <typename MdspanType, typename IdxType>
+__device__ auto increment_indices(IdxType* indices,
+                                  MdspanType const& md,
+                                  IdxType const* old_indices,
+                                  IdxType const* index_strides,
+                                  IdxType increment)
+{
+#pragma unroll
+  for (auto i = typename MdspanType::extents_type::rank_type{}; i < md.rank(); ++i) {
+    increment += index_strides[i] * old_indices[i];
+  }
+
+#pragma unroll
+  for (auto i = typename MdspanType::extents_type::rank_type{}; i < md.rank(); ++i) {
+    // Iterate through dimensions in order from slowest to fastest varying
+    auto const real_index = [](auto ind) {
+      if constexpr (std::is_same_v<typename MdspanType::layout_type, layout_f_contiguous>) {
+        return MdspanType::rank() - ind - 1;
+      } else {
+        return ind;
+      }
+    }(i);
+
+    auto cur_index = IdxType{};
+
+    while (cur_index < md.extent(real_index) - 1 && increment >= index_strides[real_index]) {
+      increment -= index_strides[real_index];
+      ++cur_index;
+    }
+    indices[real_index] = cur_index;
+  }
+
+  return increment == IdxType{};
+}
+
+/*
+ * WARNING: This kernel _must_ be launched with mdspan_copy_tile_dim x
+ * mdspan_copy_tile_dim threads per block. This restriction allows for
+ * additional optimizations at the expense of generalized launch
+ * parameters.
+ */
+template <typename DstType, typename SrcType>
+__global__ mdspan_copyable_with_kernel_t<DstType, SrcType> mdspan_copy_kernel(DstType dst,
+                                                                              SrcType src)
+{
+  using config = mdspan_copyable<true, DstType, SrcType>;
+
+  // An intermediate storage location for the data to be copied.
+  __shared__ typename config::dst_value_type tile[mdspan_copy_tile_dim][mdspan_copy_tile_dim + 1];
+
+  // Compute the cumulative product of extents in order from fastest to
```

Comment:
Yep! That general approach is how we would iterate in something akin to numpy's ravel. As a matter of fact, I believe a C++17 version of that is pretty close to how RAFT's current ravel works. The goal of this is to (where possible) read from the source in memory-layout order and write to the destination in memory-layout order. This us allows us to perform a coalesced read from global memory and coalesced write to global memory.

We won't always be able to get coalesced reads/writes with the present approach, but we will in the most common cases used by RAFT. Sometimes it will not be possible to get coalesced reads/writes at all, and then this approach will at least be correct even if it is slow.